core/hyphenator-liang.lua

```diff
@@ -114,18 +114,44 @@ local hyphenateNode = function (node)
     local hyphen = SILE.shaper:createNnodes(SILE.settings:get("font.hyphenchar"), node.options)
     local newnodes = {}
     for j, segment in ipairs(segments) do
-      local leadingApostrophe
+      local specificDiscretionary
       if segment == "" then
         SU.dump({ j, segments })
         SU.error("No hyphenation segment should ever be empty", true)
       end
       if node.options.language == "tr" then
+        -- leading apostrophe (on next segment) cancels when hyphenated
         local nextApostrophe = j < #segments and luautf8.match(segments[j+1], "^['’]")
         if nextApostrophe then
           segments[j+1] = luautf8.gsub(segments[j+1], "^['’]", "")
           local replacement = SILE.shaper:createNnodes(nextApostrophe, node.options)
-          leadingApostrophe = SILE.nodefactory.discretionary({ replacement = replacement, prebreak = hyphen })
-          leadingApostrophe.parent = node
+          specificDiscretionary = SILE.nodefactory.discretionary({ replacement = replacement, prebreak = hyphen })
+        end
+      elseif node.options.language == "ca" then
```

Comment:
So we now have another special case, similar to Turkish...
I would suggest we discuss moving those in some dedicated language-specific classes for handling post-segmentation exceptions. Other noteworthy points
- We should likely cache those discretionary (here below and for Turkish as well), instead of rebuilding them each time they are needed.
- Specific files for these hacks would need to use a language-base filename rather that merely comparing the language string. When supporting qualified BCP47 languages, we may want to distinguish `ca` (canonical Catalan = `ca-ES`) from `ca-FR` (as used in the French Pyrenees)...

Such a refactor goes beyond this bug fix, and I would suggest discussing it in a dedicated refactoring issue.

Reply:
Yes agreed. We let the Turkish happen ad hoc and I think its okay to get these in that way too, but now that we have 3+ exceptional cases it's time to review them and consider how to refactor it better.

Reply:
Linking to #1631 for cross-referencing.

Reply:
Tracked as #1966